File: node/job_dispatcher/worker.js
import DynamoDBService from "../services/dynamodb_service.js"
import Metrics from "../common/metrics.js";

class Worker{
  constructor(partitions){
    this.partitions = partitions
    this.metrics = new Metrics(Date.now(), partitions)
    this.ddbService = new DynamoDBService(this.metrics);
    this.scanTimes = new Map;
  }
  // 서비스 SQS에 넣기
  async dispatchToDestination(schedule){
    const message = `
    == PUSH JOB TO DESTINATION QUEUE ==
    QUEUE: ${schedule.jobSpec.queueName}
    JOB: ${schedule.jobSpec.jobClass}
    PARAMS: ${schedule.jobSpec.jobParams}

    `
    console.log(message);
    return this.ddbService.deleteDispatchedJob(schedule);
  }

  async dispatchOverdue(partition){
    const start = Date.now()
    return new Promise((resolve, reject)=>{
      this.ddbService
      .getOverdueJobs(partition)
      .then(({schedules, shouldImmediatelyQueryAgain})=>{
        const promises = schedules.map(async (schedule)=>{
          return this.ddbService
          .updateStatus(schedule, 'SCHEDULED', 'ACQUIRED')
          .then(schedule => this.dispatchToDestination.call(this, schedule))
        })
        Promise.all(promises)
        .then(()=>{
          console.log("all schedule dispatch promise done..")
          this.metrics.dispatchOverdue(Date.now()-start, partition, schedules.length)
          resolve(shouldImmediatelyQueryAgain);
        })
      })
      .catch((err)=>{
        console.log(err);
        resolve(true);
      })
    })
  }

  async scanGroup(){
    const start = Date.now()
    let noDelay = false;

    for(const partition of this.partitions){
      const now = Date.now()

      if(this.scanTimes.get(partition) <= now){
        const thisNoDelay = 
        await this.dispatchOverdue(partition)
        .then(scheduleImmediate=>{
          const nextScan = scheduleImmediate ? Date.now() : Date.now() + 1000
          this.scanTimes.set(partition, nextScan)
          return scheduleImmediate
        })
        noDelay = noDelay || thisNoDelay
      }else{
        noDelay = false;
      }
    }
    // TODO : post PartitionWorkerIterationEvent event
    this.metrics.scanGroup(Date.now()-start, this.partitions, noDelay)
    if(noDelay){
      // this.scanGroup(partitions) => stack overflow...
      setTimeout(()=>this.scanGroup(), 0)
    }else{
      setTimeout(()=>this.scanGroup(), 1000)
    }
  }

  // start worker thread (entry point)
  start(){
    const currentTimestamp = Date.now()
    this.partitions.forEach(partition=>{
      this.scanTimes.set(partition, currentTimestamp)
    })
    this.scanGroup()
  }
}

export default Worker

File: ruby-seq/worker.rb
require_relative './dynamodb_service'
require_relative './metrics'

class Worker
  def initialize(partitions)
    @metrics = Metrics.new(partitions)
    @partitions = partitions
    @ddbsvc = DynamoDBService.new(@metrics)
    @scan_times = Hash.new
    current_time = Time.now
    @partitions.each do |partition|
      @scan_times[partition] = current_time
    end
  end

  def dispatch_to_destination(schedule)
    msg = 
    puts "== PUSH JOB TO DESTINATION QUEUE =="
    puts "QUEUE: #{schedule.job_spec['queueName']}"
    puts "JOB: #{schedule.job_spec['jobClass']}"
    puts "QUEUE: #{schedule.job_spec['jobParams']}"
    puts ''
    @ddbsvc.delete_dispatched_jobs(schedule)
  end

  def dispatch_overdue(partition)
    start = Time.now
    query_result = @ddbsvc.get_overdue_jobs(partition)
    query_result[:schedules].each do |schedule|
      schedule = @ddbsvc.update_status(schedule, 'SCHEDULED', 'ACQUIRED')
      dispatch_to_destination(schedule)
    end
    @metrics.dispatch_overdue(Time.now-start, partition, query_result[:schedules].count)
    query_result[:should_immediately_query_again]
  end

  def scan_group
    start = Time.now
    puts 'start scanning...'
    no_delay = false
    @partitions.each do |partition|
      now = Time.now
      if @scan_times[partition] <= now
        this_no_delay = dispatch_overdue(partition)
        next_scan = this_no_delay ? Time.now : Time.now + 1
        @scan_times[partition] = next_scan
      else
        no_delay = false
      end
    end

    @metrics.scan_group(Time.now-start, no_delay)
    sleep(1) unless no_delay
    scan_group
  end
end